submissions/LuckyDnepr/task-js-oop/app.js

```diff
@@ -0,0 +1,173 @@
+    }
+}
+
+class Player extends Unit {
+    constructor (coordX, coordY, avatar) {
+        super(coordX, coordY, avatar);
+        this._freeze = true; //the player does not respond to the keyboard if _freeze === true
+    }
+    freezing() {
+        this._freeze = true; //player freezed until the win/lose banner showing
+    }
+    unfreezing() {
+        this._freeze = false; //player unfreezed when gameplay start
+    }
+    movePlayerToStart() { //randomize player start position after win/lose
+        this.x = Math.floor(Math.random() * 4) * blockWidth;
+        this.y = (Math.random() > 0.5) ? (blockHeight * 5 - 10) : (blockHeight * 4 - 10);
+    }
+    handleInput(key) {
+        if (!this._freeze) {
+            switch (key) { //choose direction depending on the key pressed
+                case "ArrowLeft":
+                    this.stepLeft();
+                    break;
+                case "ArrowRight":
+                    this.stepRight();
+                    break;
+                case "ArrowUp":
+                    this.stepUp();
+                    break;
+                case "ArrowDown":
+                    this.stepDown();
+                    break;
+                default:
+                    break;
+            }
+            this.checkWin(); //check win after move player
+        }
+    }
+    stepLeft () { //player do step left
+        if (this.x > 0) {
+            this.x -= blockWidth;
+        }
+    }
+    stepRight () { //player do step right
+        if (this.x < fieldWidth - 1.5 * blockWidth) {
+            this.x += blockWidth;
+        }
+    }
+    stepUp () { //player do step up
+        if (this.y > 0) {
+            this.y -= blockHeight;
+        }
+    }
+    stepDown () { //player do step down
+        if (this.y < fieldHeight - 1.5 * blockHeight) {
+            this.y += blockHeight;
+        }
+    }
+    checkWin () { //player on the water => you win
+        if (this.y < 0) { 
+            this._freeze = true;
+            endGame(true); //win the game
+        }
+    }
+}
+
+function endGame(win) { //end gameset
+    if (win) {
+        winPoints += 1;
+        setTimeout(() => restartGame(true), 400);
+        updatePoints();
+    } else {
+        failPoints += 1;
+        setTimeout(() => restartGame(false), 0);
+        updatePoints();
+    }
+}
+
+function restartGame(winLose) { //restart gameset - winLose - true->win; false->fail
+    winLoseBanners (winLose);
+    player.freezing();
+    player.movePlayerToStart();
+    setTimeout(() => {
+        winLoseBanners ();
+        player.unfreezing();
+    }, 750);
+}
+
+function generateEnemies(n) { //n - enemies per track - this is equivalent difficulty
+    let enemies = [];
+    for (let i = 1; i <= n; i++) {
+        enemies
+            .push(...enemyTracksHeight
+                .map((trackHeight) => {
+                    return new Enemy(Math.random() * -300, trackHeight, Math.random() * 350); //randomise start point and speed
```

Comment:
If we need a comment this means the code is not expressive enough. Function names are normally descriptive.